event-exporter/pkg/exporter/exporter.go

```diff
@@ -0,0 +1,379 @@
+		return count, fmt.Errorf("cursor error: %w", err)
+	}
+
+	return count, nil
+}
+
+func (e *HealthEventsExporter) decodeBackfillEvent(cursor client.Cursor) (*pb.HealthEvent, error) {
+	var healthEventWithStatus model.HealthEventWithStatus
+	if err := cursor.Decode(&healthEventWithStatus); err != nil {
+		slog.Warn("Failed to decode event", "error", err)
+		return nil, err
+	}
+
+	if healthEventWithStatus.HealthEvent == nil {
+		slog.Debug("Skipping nil health event")
+		return nil, nil
+	}
+
+	return healthEventWithStatus.HealthEvent, nil
+}
+
+func (e *HealthEventsExporter) publishBackfillEvent(ctx context.Context, healthEvent *pb.HealthEvent) error {
+	eventTime := "unknown"
+	if healthEvent.GeneratedTimestamp != nil {
+		eventTime = healthEvent.GeneratedTimestamp.AsTime().Format(time.RFC3339)
+	}
+
+	slog.Info("Publishing backfill event",
+		"nodeName", healthEvent.NodeName,
+		"checkName", healthEvent.CheckName,
+		"generatedAt", eventTime)
+
+	if err := e.publishWithRetry(ctx, healthEvent); err != nil {
+		return fmt.Errorf("publish event: %w", err)
+	}
+
+	return nil
+}
+
+func (e *HealthEventsExporter) waitForRateLimit(
+	ctx context.Context,
+	useRateLimiting bool,
+	rateLimiter *time.Ticker,
+) error {
+	if useRateLimiting {
+		select {
+		case <-ctx.Done():
+			slog.Error("Context done", "error", ctx.Err())
+			return ctx.Err()
+		case <-rateLimiter.C:
+		}
+	} else if ctx.Err() != nil {
+		slog.Error("Context done", "error", ctx.Err())
+		return ctx.Err()
+	}
+
+	return nil
+}
+
+func (e *HealthEventsExporter) streamEvents(ctx context.Context) error {
+	slog.Info("Starting event stream")
+
+	for {
+		select {
+		case <-ctx.Done():
+			slog.Info("Context done", "error", ctx.Err())
+			return ctx.Err()
+		case healthEvent, ok := <-e.source.Events():
+			if !ok {
+				slog.Error("Event channel closed")
+				return fmt.Errorf("event channel closed")
+			}
+
+			if err := e.processEvent(ctx, healthEvent); err != nil {
+				slog.Error("Failed to process event", "error", err)
+				return fmt.Errorf("process event: %w", err)
+			}
+		}
+	}
+}
+
+func (e *HealthEventsExporter) processEvent(ctx context.Context, healthEvent client.Event) error {
+	metrics.EventsReceived.Inc()
+
+	event, err := unmarshalHealthEvent(healthEvent)
+	if err != nil {
+		slog.Warn("Failed to unmarshal event", "error", err)
+		metrics.TransformErrors.Inc()
+
+		return e.markProcessedOrFail(ctx, healthEvent.GetResumeToken())
+	}
+
+	if event == nil {
+		slog.Debug("Skipping nil health event")
+		return e.markProcessedOrFail(ctx, healthEvent.GetResumeToken())
+	}
+
+	eventTime := "unknown"
+	if event.GeneratedTimestamp != nil {
+		eventTime = event.GeneratedTimestamp.AsTime().Format(time.RFC3339)
+	}
+
+	slog.Info("Publishing stream event",
+		"nodeName", event.NodeName,
+		"checkName", event.CheckName,
+		"generatedAt", eventTime)
+
+	if err := e.publishWithRetry(ctx, event); err != nil {
+		return fmt.Errorf("publish event: %w", err)
+	}
```

Comment:
This seems to be identical to `publishBackfillEvent` expect for the logline. Since we already have a log line and metrics that says it is in the backfill state, can we refactor so that the publish flow is the same irrespective of events being backfilled or not?